provisioning/baremetal_pod_test.go

```diff
@@ -0,0 +1,214 @@
+				Value: "172.30.20.3/24",
+			},
+		},
+		{
+			name:       "Unmanaged ProvisioningInterface",
+			configName: provisioningInterface,
+			spec:       unmanagedSpec,
+			expectedEnvVar: corev1.EnvVar{
+				Name:  provisioningInterface,
+				Value: "ensp0",
+			},
+		},
+		{
+			name:       "Disabled MachineOsUrl",
+			configName: machineImageUrl,
+			spec:       disabledSpec,
+			expectedEnvVar: corev1.EnvVar{
+				Name:  machineImageUrl,
+				Value: "http://172.22.0.1/images/rhcos-44.81.202001171431.0-openstack.x86_64.qcow2.gz?sha256=e98f83a2b9d4043719664a2be75fe8134dc6ca1fdbde807996622f8cc7ecd234",
+			},
+		},
+		{
+			name:       "Disabled ProvisioningInterface",
+			configName: provisioningInterface,
+			spec:       disabledSpec,
+			expectedEnvVar: corev1.EnvVar{
+				Name:  provisioningInterface,
+				Value: "",
+			},
+		},
+	}
+	for _, tc := range tCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Logf("Testing tc : %s", tc.name)
+			actualEnvVar := buildEnvVar(tc.configName, &tc.spec)
+			assert.Equal(t, tc.expectedEnvVar, actualEnvVar, fmt.Sprintf("%s : Expected : %s Actual : %s", tc.configName, tc.expectedEnvVar, actualEnvVar))
+			return
+		})
+	}
+
+}
+
+func TestNewMetal3InitContainers(t *testing.T) {
+	managedSpec := metal3iov1alpha1.ProvisioningSpec{
+		ProvisioningInterface:     "eth0",
+		ProvisioningIP:            "172.30.20.3",
+		ProvisioningNetworkCIDR:   "172.30.20.0/24",
+		ProvisioningDHCPRange:     "172.30.20.11, 172.30.20.101",
+		ProvisioningOSDownloadURL: "http://172.22.0.1/images/rhcos-44.81.202001171431.0-openstack.x86_64.qcow2.gz?sha256=e98f83a2b9d4043719664a2be75fe8134dc6ca1fdbde807996622f8cc7ecd234",
+		ProvisioningNetwork:       "Managed",
+	}
+	images := Images{
+		BaremetalOperator:   expectedBaremetalOperator,
+		Ironic:              expectedIronic,
+		IronicInspector:     expectedIronicInspector,
+		IpaDownloader:       expectedIronicIpaDownloader,
+		MachineOsDownloader: expectedMachineOsDownloader,
+		StaticIpManager:     expectedIronicStaticIpManager,
+	}
+	tCases := []struct {
+		name               string
+		expectedContainers []corev1.Container
+	}{
+		{
+			name: "valid config",
+			expectedContainers: []corev1.Container{
+				{
+					Name:  "metal3-ipa-downloader",
+					Image: images.IpaDownloader,
+				},
+				{
+					Name:  "metal3-machine-os-downloader",
+					Image: images.MachineOsDownloader,
+				},
+				{
+					Name:  "metal3-static-ip-set",
+					Image: images.StaticIpManager,
+				},
+			},
+		},
+	}
+	for _, tc := range tCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Logf("Testing tc : %s", tc.name)
+			actualContainers := newMetal3InitContainers(&images, &managedSpec)
+			assert.Equal(t, len(tc.expectedContainers), len(actualContainers), fmt.Sprintf("%s : Expected number of Init Containers : %d Actual number of Init Containers : %d", tc.name, len(tc.expectedContainers), len(actualContainers)))
+		})
+	}
+
+}
+
+func TestNewMetal3Containers(t *testing.T) {
```

Comment:
I think that such test (and from a broader point of view the one on the Deployment, not yet present) deserves all the possible attentions, since it represents a core feature of the operator: an error in one of the generated fields could potentially compromise the entire metal3 deployment. 
So, from this point of view, it could be worth to "enrich" such kind of tests by taking care that all the fields contain the expected value - instead of just counting the number of containers. Thus, it seems a perfect fit for the golden file testing approach: I've prepared a sample PR to illustrate the approach (see https://github.com/sadasu/cluster-baremetal-operator/pull/3). It could work well directly on the generated overall deployment

Reply:
Thanks for this suggestion @andfasano. I was looking for ways to better test the Deployment object. I'll use your suggest technique for the Deployment object.